### PR TITLE
[MRESOLVER-115] make checksum algorithms configurable

### PR DESCRIPTION
**Pull request by kwin**

### CODE REVIEW

**Comment by kwin** [comment]
@michael-o Hopefully I addressed all your concerns in https://github.com/apache/maven-resolver/commit/347eee2fa6f0634cd459dafe3382e7ef36c87f19

**Comment by michael-o** [comment]
`~/apache-maven-3.7.0-SNAPSHOT/bin/mvn deploy -Daether.checksums.algorithms=SHA-256 -Dorg.slf4j.simpleLogger.log.org.apache.maven.wagon.providers.http.httpclient=debug -Dorg.slf4j.simpleLogger.log.org.apache.maven.wagon.providers.http.httpclient.wire=off` does not yet work. All checksums are uploaded.

**Comment by kwin** [comment]
How will the CLI argument `-Daether.checksums.algorithms=SHA-256` ever end up in https://github.com/apache/maven-resolver/blob/47edcfe69c4e52ced4cb93d65b7348b5645cdd68/maven-resolver-api/src/main/java/org/eclipse/aether/DefaultRepositorySystemSession.java#L519? I couldn't find the code setting the configProperties...

**Comment by michael-o** [comment]
OK, I have found the reason. It cannot work. `org.apache.maven.internal.aether.DefaultRepositorySystemSessionFactory` sets only simple strings. `ConfigUtils.getList()` expects the object to be a collection. This means that `#getList() needs an if else for string and perform a split.

**Comment by kwin** [comment]
@michael-o Shouldn't the fix rather be in https://github.com/apache/maven/blob/a2b800de32cdb9adc1e64a43a0fc32e3ba878103/maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java#L105? IMHO parsing and splitting strings should not be done inside Maven Resolver.

**Comment by michael-o** [comment]
At glance yes, but this would require you to know the semantics of each and every property. This is not possible. They are completely opaque to Maven and their meaning is only known to its consumer.  If you look at `#getBoolean()` it does parse the value too. @rfscholte What is your opinion?

**Comment by rfscholte** [comment]
First thought: add a specific property for it to `ConfigurationProperties` and make sure DefaultRepositorySystemSessionFactory sets it properly. And yes, that implies it will only work with the next Maven release.

**Comment by michael-o** [comment]
> 
> 
> First thought: add a specific property for it to `ConfigurationProperties` and make sure DefaultRepositorySystemSessionFactory sets it properly. And yes, that implies it will only work with the next Maven release.

This feels to specific. The only cases Maven uses this class where `MavenExecutionRequest` has distinct values for this and yet they can still be overridden. We'd bascially need to get this specific value from system and user properties, process and put back again. Very, very hacky....from my POV.

**Comment by michael-o** [comment]
> 
> 
> How will the CLI argument `-Daether.checksums.algorithms=SHA-256` ever end up in
> 
> https://github.com/apache/maven-resolver/blob/47edcfe69c4e52ced4cb93d65b7348b5645cdd68/maven-resolver-api/src/main/java/org/eclipse/aether/DefaultRepositorySystemSession.java#L519
> ? I couldn't find the code setting the configProperties...

Via org.apache.maven.internal.aether.DefaultRepositorySystemSessionFactory.

**Comment by kwin** [comment]
Another option is to rely on String for this option and do the splitting afterwards like in https://github.com/apache/maven-resolver/blob/47edcfe69c4e52ced4cb93d65b7348b5645cdd68/maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/SslConfig.java#L58

**Comment by kwin** [comment]
@michael-o Please try again with https://github.com/apache/maven-resolver/pull/53/commits/1762c3df96e6c2f5901cde8c08e5f17034c415b3

**Comment by michael-o** [comment]
First upload tests look fine. I am still running Nexus 2.14.16 at work. Will try download and Core ITs. If nothing fails, I will create an upstream branch for this and leave a couple of days of review for others. When no objections are expressed, I will merge by Tue.

**Comment by michael-o** [comment]
Branch pushed, local tests are positive.

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java [hunk 1]
```diff
@@ -155,10 +168,12 @@ public URI getLocation( Metadata metadata, boolean upload )
 
         private List<Checksum> getChecksums( URI location )
         {
-            return Arrays.asList( Checksum.forLocation( location, "SHA-512" ),
-                                  Checksum.forLocation( location, "SHA-256" ),
-                                  Checksum.forLocation( location, "SHA-1" ),
-                                  Checksum.forLocation( location, "MD5" ) );
+            List<Checksum> checksums = new LinkedList<>( );
```
Comment (elharo) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java:171]: nit: arraylist is almost always better, data structures textbooks to the contrary.

Comment (michael-o) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/Maven2RepositoryLayoutFactory.java:171]: Agree, does this matter for a bunch of elements? Less than 10?

**Comment by michael-o** [comment]
Haven't received any real objections. Will go on with the merge.

**Comment by jira-importer** [comment]
Resolve #863

**Comment by jira-importer** [comment]
Resolve #863

**Pull request closed without merging**